plugins/org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/ui/contentAssist/ContentAssistFragment2.xtend

```diff
@@ -0,0 +1,330 @@
+		fileAccessFactory.createXtendFile(grammar.proposalProviderClass, '''
+			/**
+			 * See https://www.eclipse.org/Xtext/documentation/304_ide_concepts.html#content-assist
+			 * on how to customize the content assistant.
+			 */
+			class �grammar.proposalProviderClass.simpleName� extends �grammar.genProposalProviderClass� {
+			}
+		''').writeTo(projectConfig.eclipsePluginSrc)
+	}
+
+	def generateJavaProposalProviderStub() {
+		fileAccessFactory.createJavaFile(grammar.proposalProviderClass, '''
+			/**
+			 * See https://www.eclipse.org/Xtext/documentation/304_ide_concepts.html#content-assist
+			 * on how to customize the content assistant.
+			 */
+			public class �grammar.proposalProviderClass.simpleName� extends �grammar.genProposalProviderClass� {
+			}
+		''').writeTo(projectConfig.eclipsePluginSrc)
+	}
+
+	// generation of the 'Abstract...ProposalProvider'
+
+	
+	def generateGenJavaProposalProvider() {
+		// excluded features are those that stem from inherited grammars,
+		//  they are handled by the super grammars' proposal provider
+		val excludedFqnFeatureNames = grammar.getFQFeatureNamesToExclude
+		val processedNames = newHashSet()
+
+		// determine all assignments within the grammar that are not excluded
+		//  and not handled yet, (iterable is evaluated lazily!)
+		val assignments = grammar.containedAssignments().fold(<Assignment>newArrayList()) [ candidates, assignment |
+			val fqFeatureName = assignment.FQFeatureName
+			if (!processedNames.contains(fqFeatureName) && !excludedFqnFeatureNames.contains(fqFeatureName)) {
+				processedNames += fqFeatureName;
+				candidates += assignment;
+			}
+			candidates
+		]
+		
+		// determine the remaining rules that are not excluded
+		//  and not handled yet, (iterable is evaluated lazily!)
+		val remainingRules = grammar.rules.fold(<AbstractRule>newArrayList()) [candidates, rule |
+  			val fqnFeatureName = rule.FQFeatureName
+			if (!processedNames.contains(fqnFeatureName) && !excludedFqnFeatureNames.contains(fqnFeatureName)) {
+				processedNames += fqnFeatureName
+				candidates += rule
+			}
+			candidates
+		]
+		
+		val superClass = grammar.getGenProposalProviderSuperClass
+		fileAccessFactory.createJavaFile(grammar.getGenProposalProviderClass, '''
+			/**
+			 * Represents a generated, default implementation of superclass {@link �superClass�}.
+			 * Methods are dynamically dispatched on the first parameter, i.e., you can override them 
+			 * with a more concrete subtype. 
+			 */
+			@SuppressWarnings("all")
+			public class �grammar.getGenProposalProviderClass.simpleName� extends �superClass� {
+			
+				�FOR assignment : assignments�
+					�assignment.handleAssignment�
+			  	�ENDFOR�
+				
+				�FOR rule : remainingRules�
+					public void complete�rule.FQFeatureName�(�EObject� model, �RuleCall� ruleCall, �
+							contentAssistContextClass� context, �ICompletionProposalAcceptorClass� acceptor) {
+						// subclasses may override
+					}
+		    	�ENDFOR�
```

Comment:
the line ``public void complete«r...``was quite long so I inserted a line break before the ``contentAssistContextClass`` method call; this way the indentation in the generated code is not affected